be/src/olap/rowset/segment_v2/segment_writer.cpp

```diff
@@ -0,0 +1,205 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "olap/rowset/segment_v2/segment_writer.h"
+
+#include "env/env.h" // Env
+#include "olap/row_block.h" // RowBlock
+#include "olap/row_cursor.h" // RowCursor
+#include "olap/rowset/segment_v2/column_writer.h" // ColumnWriter
+#include "olap/short_key_index.h"
+
+namespace doris {
+namespace segment_v2 {
+
+const char* k_segment_magic = "D0R1";
+const uint32_t k_segment_magic_length = 4;
+
+SegmentWriter::SegmentWriter(
+        std::string fname, const std::vector<ColumnSchemaV2>& schema,
+        size_t num_short_keys, uint32_t segment_id,
+        uint32_t num_rows_per_block, bool delete_flag)
+    : _fname(std::move(fname)),
+    _schema(schema),
+    _num_short_keys(num_short_keys),
+    _segment_id(segment_id),
+    _num_rows_per_block(num_rows_per_block),
+    _delete_flag(delete_flag) {
+}
+
+SegmentWriter::~SegmentWriter() {
+    for (auto writer : _column_writers) {
+        delete writer;
+    }
+}
+
+Status SegmentWriter::init(uint32_t write_mbytes_per_sec) {
+    // create for write
+    RETURN_IF_ERROR(Env::Default()->new_writable_file(_fname, &_output_file));
+
+    for (size_t cid = 0; cid < _schema.size(); ++cid) {
+        auto& column_schema = _schema[cid];
+
+        ColumnMetaPB* column_meta = _footer.add_columns();
+        column_meta->set_column_id(cid);
+        column_meta->set_unique_id(column_schema.field_info().unique_id);
+        bool is_nullable = column_schema.field_info().is_allow_null;
+        column_meta->set_is_nullable(is_nullable);
+
+        // 
+        const TypeInfo* type_info = column_schema.type_info();
+
+        ColumnWriterOptions opts;
+        std::unique_ptr<ColumnWriter> writer(new ColumnWriter(opts, type_info, is_nullable, _output_file.get()));
+        RETURN_IF_ERROR(writer->init());
+        _column_writers.push_back(writer.release());
+    }
+
+    std::vector<ColumnSchemaV2> short_keys;
+    for (size_t i = 0; i < _num_short_keys; ++i) {
+        short_keys.push_back(_schema[i]);
+    }
+    _index_builder.reset(
+        new ShortKeyIndexBuilder(short_keys, _segment_id, _num_rows_per_block, _delete_flag));
+    RETURN_IF_ERROR(_index_builder->init());
+    return Status::OK();
+}
+
+Status SegmentWriter::write_batch(RowBlock* block, RowCursor* cursor, bool is_finalize) {
```

Comment:
I think the cursor is a helper argument, can be not a argument of this function.
I may be a class member or a tmp variable

Reply:
Yes, this will be changed in next patch